src/kOS/UserIO/TelnetSingletonServer.cs

```diff
@@ -170,15 +170,26 @@ public TelnetSingletonServer(TelnetMainServer mainServer, TcpClient client, int
             rawStream = client.GetStream();
             lock (keepAliveAccess)
                 gotSomeRecentTraffic = true;
-            inThread = new Thread(DoInThread);
-            outThread = new Thread(DoOutThread);
             outQueue = new Queue<char>();
             inQueue = new Queue<char>();
+            SpawnWelcomeMenu();
+            inThread = new Thread(DoInThread);
+            outThread = new Thread(DoOutThread);
```

Comment:
This is the core change that in turn caused all the other changes.

The SpawnWelcomeMenu() is now happening before the threads are forked off, back in the main thread.
Previously, it was done inside ``outThread``'s ``ContinuousChecks()``.

To make this work, it was necessary to move from a design of "throw TelnetWelcomeMenu away and make a new one each time you go back to the menu screen", to one of "re-use the same TelnetWelcomeMenu and just reset it."

That change is the primary reason for most of the rest of what's in this commit.